roles/swlib/tasks/gcscopy.yml

```diff
@@ -17,177 +17,96 @@
-  failed_when: shell_result.stderr != ""
+  changed_when: "'Date    Time    Name' not in shell_result.stdout"
+  delegate_to: 127.0.0.1
 
-- name: gcscopy | Copy GI patch software from GCS to target instance
+- name: gcscopy | Copy GI patches from GCS to target instance
   shell: |
     set -o pipefail
-    ssh -i {{ ansible_ssh_private_key_file }} {{ ansible_ssh_extra_args }} {{ ansible_ssh_user }}@{{ ansible_ssh_host }} \
-      "ls {{ swlib_path }}/{{ item.patchfile }}"
-    if [[ $? -eq 0 ]]
+    if ! ssh -i {{ ansible_ssh_private_key_file }} {{ ansible_ssh_extra_args }} {{ ansible_ssh_user }}@{{ ansible_ssh_host }} \
+      "unzip -l {{ swlib_path }}/{{ item.patchfile }}"
       then
-        ssh -i {{ ansible_ssh_private_key_file }} {{ ansible_ssh_extra_args }} {{ ansible_ssh_user }}@{{ ansible_ssh_host }} \
-          "unzip -t {{ swlib_path }}/{{ item.patchfile }}"
-        if [[ $? -ne 0 ]]
-          then
-            gsutil cat gs://{{ swlib_mount_src }}/{{ item.patchfile }} | ssh -i {{ ansible_ssh_private_key_file }} {{ ansible_ssh_extra_args }} \
-              {{ ansible_ssh_user }}@{{ ansible_ssh_host }} "cat > {{ swlib_path }}/{{ item.patchfile }}"
-        fi
-      else
         gsutil cat gs://{{ swlib_mount_src }}/{{ item.patchfile }} | ssh -i {{ ansible_ssh_private_key_file }} {{ ansible_ssh_extra_args }} \
           {{ ansible_ssh_user }}@{{ ansible_ssh_host }} "cat > {{ swlib_path }}/{{ item.patchfile }}"
     fi
-  args:
-    executable: /bin/bash
-  with_items:
-    - "{{ gi_patches }}"
-  when: item.release == oracle_rel and patching_type == "gi_patching"
-  delegate_to: 127.0.0.1
-
-- name: gcscopy | Verify GI patch software zipfile integrity
-  shell: |
-    set -o pipefail
-    {% for i in item.patchfile %}
-    unzip -t {{ swlib_path }}/{{ i }} > /dev/null
-        {% endfor %}
   register: shell_result
   args:
     executable: /bin/bash
   with_items:
     - "{{ gi_patches }}"
   when: item.release == oracle_rel and patching_type == "gi_patching"
-  failed_when: shell_result.stderr != ""
+  changed_when: "'Date    Time    Name' not in shell_result.stdout"
+  delegate_to: 127.0.0.1
 
 # overwrite opatch zipfile on BMS host from gcs every time that may be cached currently in {{ swlib_path }}
 # reason to overwrite is that the opatch zipfile in MOS is named the same, example: p6880880_190000_Linux-x86-64.zip
 # this change downloads OPatch from GCS bucket every time with neglible overhead as opatch zip file is typically ~120M
 - name: gcscopy | Copy GI OPatch update files from GCS to target instance
   shell: |
     set -o pipefail
-    gsutil cat gs://{{ swlib_mount_src }}/{{ item.patchfile }} | ssh -i {{ ansible_ssh_private_key_file }} {{ ansible_ssh_extra_args }} \
+    if ! ssh -i {{ ansible_ssh_private_key_file }} {{ ansible_ssh_extra_args }} {{ ansible_ssh_user }}@{{ ansible_ssh_host }} \
+      "unzip -l {{ swlib_path }}/{{ item.patchfile }}"
```

Comment:
IIUC, the logic proposed here is:
if the opatch file (say: `p6880880_190000_Linux-x86-64.zip`) exists on the server and it is intact (as borne by the rc of the `unzip -l`), then we DO NOT copy the file from GCS.

I think we need to copy the opatch files every time from GCS, whether the BM host has it or not in the swlib path on the host.

So, the `if` condition may not be needed.

Reply:
Good catch;  I thought I had different logic here, but I guess I didn't.

https://gist.github.com/mfielding/07337b39ec43a3660aea8ed47b849882